In [1]:
import numpy as np
import h5py

In [2]:
sv_data = h5py.File("SVHN_single_grey1.h5")

In [3]:
list(sv_data)

['X_test', 'X_train', 'X_val', 'y_test', 'y_train', 'y_val']

In [4]:
X_test = sv_data['X_test']
X_train = sv_data['X_train']
X_val = sv_data['X_val']
y_test = sv_data['y_test']
y_train = sv_data['y_train']
y_val = sv_data['y_val']

In [5]:
X_train.shape

(42000, 32, 32)

In [6]:
y_train.shape

(42000,)

In [7]:
X_test.shape

(18000, 32, 32)

In [8]:
y_test.shape

(18000,)

In [9]:
X_val.shape

(60000, 32, 32)

In [10]:
y_val.shape

(60000,)

In [11]:
train_X = np.reshape(X_train,(42000,32*32))

In [12]:
train_X.shape

(42000, 1024)

In [13]:
test_X = np.reshape(X_test,(18000,32*32))

In [14]:
test_X.shape

(18000, 1024)

In [15]:
from sklearn.neighbors import KNeighborsClassifier

In [16]:
model = KNeighborsClassifier(n_neighbors=15)
model.fit(train_X, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=15, p=2,
           weights='uniform')

In [17]:
y_pred = model.predict(test_X)

In [18]:
from sklearn import metrics
from sklearn.metrics import confusion_matrix

In [19]:
print("KNN Accuracy:", metrics.accuracy_score(y_pred,y_test))

KNN Accuracy: 0.5235555555555556


In [20]:
print("KNN Confusion Metrix: ", metrics.confusion_matrix(y_pred,y_test))

KNN Confusion Metrix:  [[1233   95  100  136  117  163  318   98  249  335]
 [  68 1335  215  251  247  164  122  227  115  137]
 [  39   70  992  136   41   62   38  116   62   70]
 [  39   83  107  734   60  268   72   84  119   92]
 [  51   75   55   50 1182   63  129   26  100   65]
 [  46   34   30  160   14  683  128   35  120  108]
 [ 113   32   37   42   54  156  741   35  253   56]
 [  37   51  138   54   17   31   23 1116   28   69]
 [  86   21   54   98   34  114  210   30  655  119]
 [ 102   32   75   58   46   64   51   41  111  753]]


In [21]:
cr=metrics.classification_report(y_test,y_pred)
print(cr)

             precision    recall  f1-score   support

          0       0.43      0.68      0.53      1814
          1       0.46      0.73      0.57      1828
          2       0.61      0.55      0.58      1803
          3       0.44      0.43      0.43      1719
          4       0.66      0.65      0.66      1812
          5       0.50      0.39      0.44      1768
          6       0.49      0.40      0.44      1832
          7       0.71      0.62      0.66      1808
          8       0.46      0.36      0.41      1812
          9       0.56      0.42      0.48      1804

avg / total       0.53      0.52      0.52     18000



In [22]:
import tensorflow as tf

In [23]:
#Initialize Sequential model
model = tf.keras.models.Sequential()
#Reshape data from 2D to 1D -> 32X32 to 1024
model.add(tf.keras.layers.Reshape((1024,),input_shape=(32,32)))

In [24]:
#Normalize the data
model.add(tf.keras.layers.BatchNormalization())

Instructions for updating:
Colocations handled automatically by placer.


In [25]:
#Hidden layers
model.add(tf.keras.layers.Dense(200, activation='relu'))
model.add(tf.keras.layers.Dense(100, activation='relu'))

#Output layer
model.add(tf.keras.layers.Dense(10, activation='softmax', name='Output'))

In [26]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [27]:
trainY = tf.keras.utils.to_categorical(y_train, num_classes=10)
testY = tf.keras.utils.to_categorical(y_test, num_classes=10)

In [28]:
X_train.shape

(42000, 32, 32)

In [29]:
y_train.shape

(42000,)

In [30]:
X_test.shape

(18000, 32, 32)

In [31]:
y_test.shape

(18000,)

In [32]:
X_val.shape

(60000, 32, 32)

In [33]:
y_val.shape

(60000,)

In [34]:
#Train the model
X_test_arr = np.array(X_test)
model.fit(X_train,trainY,          
          validation_data=(X_test_arr, testY),
          epochs=30,
          batch_size=100,validation_split = 0.01,
         shuffle='batch')

Train on 42000 samples, validate on 18000 samples
Instructions for updating:
Use tf.cast instead.
Epoch 1/30
42000/42000 [==============================] - 4s 96us/sample - loss: 2.0574 - acc: 0.2930 - val_loss: 1.7074 - val_acc: 0.4669
Epoch 2/30
42000/42000 [==============================] - 4s 90us/sample - loss: 1.4912 - acc: 0.5472 - val_loss: 1.2831 - val_acc: 0.6232
Epoch 3/30
42000/42000 [==============================] - 4s 89us/sample - loss: 1.1997 - acc: 0.6498 - val_loss: 1.0868 - val_acc: 0.6836
Epoch 4/30
42000/42000 [==============================] - 4s 85us/sample - loss: 1.0539 - acc: 0.6914 - val_loss: 0.9888 - val_acc: 0.7162
Epoch 5/30
42000/42000 [==============================] - 4s 85us/sample - loss: 0.9585 - acc: 0.7213 - val_loss: 0.9153 - val_acc: 0.7348
Epoch 6/30
42000/42000 [==============================] - 4s 87us/sample - loss: 0.8874 - acc: 0.7402 - val_loss: 0.8629 - val_acc: 0.7500
Epoch 7/30
42000/42000 [==============================] - 4s 85us/sa

In [35]:
loss_and_metrics = model.evaluate(X_test, testY)
print(loss_and_metrics)

18000/18000 [==============================] - 1s 78us/sample - loss: 0.5955 - acc: 0.8350
[0.595527134709888, 0.835]


In [36]:
y_pred=model.predict_classes(X_test)

In [38]:
cr=metrics.classification_report(y_test,y_pred)
print(cr)

             precision    recall  f1-score   support

          0       0.86      0.88      0.87      1814
          1       0.84      0.85      0.85      1828
          2       0.89      0.85      0.86      1803
          3       0.81      0.78      0.79      1719
          4       0.85      0.87      0.86      1812
          5       0.81      0.83      0.82      1768
          6       0.85      0.79      0.82      1832
          7       0.87      0.87      0.87      1808
          8       0.78      0.79      0.79      1812
          9       0.80      0.83      0.81      1804

avg / total       0.84      0.83      0.83     18000



## The F1-Score obtained as 0.52 on KNN and 0.83 on Neural networks. Neural Networks works better than KNN. 